pkg/reconciler/identity/resources/service_account.go

```diff
@@ -17,29 +17,30 @@ limitations under the License.
 package resources
 
 import (
+	"fmt"
 	"strings"
 
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 const (
-	workloadIdentityKey = "iam.gke.io/gcp-service-account"
+	WorkloadIdentityKey = "iam.gke.io/gcp-service-account"
 )
 
 // GenerateServiceAccountName generates a k8s ServiceAccount name according to GCP ServiceAccount
-func GenerateServiceAccountName(gServiceAccount string) string {
-	return strings.Split(gServiceAccount, "@")[0]
+func GenerateServiceAccountName(gServiceAccount, clusterName string) string {
+	return fmt.Sprintf("%s-%s", strings.Split(gServiceAccount, "@")[0], clusterName)
```

Comment:
What happens to existing resources when this ServiceAccount name changes?

Reply:
Existing resources without the clusterName annotation will have error in Identity Reconciler:
```
 if clusterName == "" {
		err := fmt.Errorf(`not able to get cluster name, please provide it by adding annotation "cluster-name=$CLUSTER_NAME" to source`)
		status.MarkWorkloadIdentityFailed(identifiable.ConditionSet(), workloadIdentityFailed, err.Error())
		return nil, fmt.Errorf("failed to get cluster name: %w", err)
	}
```
but thinking together with a previous comment about leaving the clusterName annotation blank if user disable metadata, maybe I can allow the defaulting webhook to leave the annotation blank even if it is onGKE, and use the util.getClusterName to get clusterName again in Identity Reconciler, so that existing resources will get they clusterName even the annotation is not populated by the defaulting webhook.

Reply:
And those users who disabled metadata access still need to provide the cluster name ....

Reply:
We set the spec.googleServiceAccount is immutable, so before this change, the ServiceAccount name will not change. Yes, we need a immutable check also for the clusterName annotation.